src/SSHDebugPS/PipeConnection.cs

```diff
@@ -0,0 +1,251 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using Microsoft.VisualStudio.Debugger.Interop.UnixPortSupplier;
+
+namespace Microsoft.SSHDebugPS
+{
+    internal abstract class PipeConnection : Connection
+    {
+        private readonly object _lock = new object();
+        private readonly IPipeTransportSettings _settings;
+        private readonly Connection _outerConnection;
+
+        private readonly ShellExecutionManager _shellExecutionManager;
+        private readonly LinkedList<IRawShell> _shellList = new LinkedList<IRawShell>();
+        private bool _isClosed;
+        private string _name;
+
+        public override string Name => _name;
+
+        protected Connection OuterConnection => _outerConnection;
+        protected IPipeTransportSettings TransportSettings => _settings;
+        protected bool IsClosed => _isClosed;
+
+        /// <summary>
+        /// Create a new pipe connection object
+        /// </summary>
+        /// <param name="pipeTransportSettings">Settings</param>
+        /// <param name="outerConnection">[Optional] the SSH connection (or maybe something else in future) used to connect to the target.</param>
+        /// <param name="name">The full name of this connection</param>
+        /// <param name="container">The name of the container. For local, this is the same as 'name'.</param>
+        public PipeConnection(IPipeTransportSettings pipeTransportSettings, Connection outerConnection, string name, int? timeout)
+        {
+            Debug.Assert(pipeTransportSettings != null);
+            Debug.Assert(!string.IsNullOrEmpty(name));
+            //Debug.Assert(!string.IsNullOrEmpty(container));
+
+            _name = name;
+            _settings = pipeTransportSettings;
+            if (outerConnection != null)
+            {
+                _outerConnection = outerConnection;
+                DefaultTimeout = 15000;
+            }
+            else
+            {
+                DefaultTimeout = timeout.HasValue ? timeout.Value : 5000;
+            }
+
+            _shellExecutionManager = new ShellExecutionManager(CreateShellFromSettings(_settings, _outerConnection));
+        }
+
+        public override void BeginExecuteAsyncCommand(string commandText, bool runInShell, IDebugUnixShellCommandCallback callback, out IDebugUnixShellAsyncCommand asyncCommand)
+        {
+            throw new NotImplementedException();
```

Comment:
>throw new NotImplementedException(); [](start = 12, length = 36)

Is there a reason we override this just to throw NotImplementedException? Should we add a comment?

Reply:
I think it was left from when i had it non-abstract. i'll remove it.